services/alert/api.go

```diff
@@ -0,0 +1,503 @@
+package alert
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"path"
+	"sort"
+	"strings"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/influxdata/kapacitor/alert"
+	client "github.com/influxdata/kapacitor/client/v1"
+	"github.com/influxdata/kapacitor/services/httpd"
+)
+
+const (
+	alertsPath         = "/alerts"
+	alertsPathAnchored = "/alerts/"
+
+	topicsPath             = alertsPath + "/topics"
+	topicsPathAnchored     = alertsPath + "/topics/"
+	topicsBasePath         = httpd.BasePreviewPath + topicsPath
+	topicsBasePathAnchored = httpd.BasePreviewPath + topicsPathAnchored
+
+	topicEventsPath   = "events"
+	topicHandlersPath = "handlers"
+
+	eventsPattern   = "*/" + topicEventsPath
+	eventPattern    = "*/" + topicEventsPath + "/*"
+	handlersPattern = "*/" + topicHandlersPath
+	handlerPattern  = "*/" + topicHandlersPath + "/*"
+
+	eventsRelation   = "events"
+	handlersRelation = "handlers"
+)
+
+type apiServer struct {
+	Registrar    HandlerSpecRegistrar
+	Topics       Topics
+	Persister    TopicPersister
+	routes       []httpd.Route
+	HTTPDService interface {
+		AddPreviewRoutes([]httpd.Route) error
+		DelRoutes([]httpd.Route)
+	}
+	logger *log.Logger
+}
+
+func (s *apiServer) Open() error {
+	// Define API routes
+	s.routes = []httpd.Route{
+		{
+			Method:      "GET",
+			Pattern:     topicsPath,
+			HandlerFunc: s.handleListTopics,
+		},
+		{
+			Method:      "GET",
+			Pattern:     topicsPathAnchored,
+			HandlerFunc: s.handleRouteTopicGet,
+		},
+		{
+			Method:      "POST",
+			Pattern:     topicsPathAnchored,
+			HandlerFunc: s.handleRouteTopicPost,
+		},
+		{
+			Method:      "PATCH",
+			Pattern:     topicsPathAnchored,
+			HandlerFunc: s.handleRouteTopicPatch,
+		},
+		{
+			Method:      "PUT",
+			Pattern:     topicsPathAnchored,
+			HandlerFunc: s.handleRouteTopicPut,
+		},
+		{
+			Method:      "DELETE",
+			Pattern:     topicsPathAnchored,
+			HandlerFunc: s.handleRouteTopicDelete,
+		},
+		{
+			// Satisfy CORS checks.
+			Method:      "OPTIONS",
+			Pattern:     topicsPathAnchored,
+			HandlerFunc: httpd.ServeOptions,
+		},
+	}
+
+	return s.HTTPDService.AddPreviewRoutes(s.routes)
+}
+
+func (s *apiServer) Close() error {
+	if s.HTTPDService != nil {
+		s.HTTPDService.DelRoutes(s.routes)
+	}
+	return nil
+}
+
+type sortedTopics []client.Topic
+
+func (s sortedTopics) Len() int               { return len(s) }
+func (s sortedTopics) Less(i int, j int) bool { return s[i].ID < s[j].ID }
+func (s sortedTopics) Swap(i int, j int)      { s[i], s[j] = s[j], s[i] }
+
+func (s *apiServer) handleListTopics(w http.ResponseWriter, r *http.Request) {
+	pattern := r.URL.Query().Get("pattern")
+	if err := validatePattern(pattern); err != nil {
+		httpd.HttpError(w, fmt.Sprint("invalid pattern: ", err.Error()), true, http.StatusBadRequest)
+		return
+	}
+	minLevelStr := r.URL.Query().Get("min-level")
```

Comment:
Not a really an issue since this isn't a hot code path, but `URL.Query()` generates a new map each time its called.

https://golang.org/src/net/url/url.go?s=27419:27447#L967